File: pymd/backend/app/schemas/environment.py
"""Environment schemas"""
from datetime import datetime
from typing import Optional, List
from uuid import UUID
from pydantic import BaseModel, Field


class EnvironmentBase(BaseModel):
    """Base environment schema"""

    name: str = Field(..., min_length=1, max_length=100, description="Environment name")
    python_version: str = Field(default="3.11", pattern="^3\\.1[0-9]$", description="Python version")


class EnvironmentCreate(EnvironmentBase):
    """Environment creation schema"""

    pass


class EnvironmentUpdate(BaseModel):
    """Environment update schema"""

    name: Optional[str] = Field(None, min_length=1, max_length=100)
    python_version: Optional[str] = Field(None, pattern="^3\\.1[0-9]$")


class EnvironmentResponse(BaseModel):
    """Environment response schema"""

    id: UUID
    user_id: UUID
    name: str
    python_version: str
    volume_name: str
    status: str
    total_size_bytes: int
    package_count: int
    last_used_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class EnvironmentListResponse(BaseModel):
    """Environment list response"""

    environments: List[EnvironmentResponse]
    total: int


class EnvironmentStatsResponse(BaseModel):
    """Environment statistics response"""

    environment_count: int
    environment_limit: int
    total_size_bytes: int
    size_limit_bytes: int
    total_packages: int
    package_limit: int
    can_create_environment: bool


class PackageBase(BaseModel):
    """Base package schema"""

    package_name: str = Field(..., min_length=1, max_length=255, description="Package name")
    version: Optional[str] = Field(None, max_length=100, description="Package version")


class PackageInstallRequest(BaseModel):
    """Package installation request"""

    packages: List[str] = Field(..., min_items=1, max_items=10, description="List of packages to install")


class PackageResponse(BaseModel):
    """Package response schema"""

    id: UUID
    environment_id: UUID
    package_name: str
    version: Optional[str]
    size_bytes: Optional[int]
    installed_at: datetime

    model_config = {"from_attributes": True}


class PackageListResponse(BaseModel):
    """Package list response"""

    packages: List[PackageResponse]
    total: int
    total_size_bytes: int


class ExecutionRequest(BaseModel):
    """Code execution request"""

    code: str = Field(..., min_length=1, description="Python code to execute")
    environment_id: Optional[UUID] = Field(None, description="Environment ID (uses default if not provided)")


class ExecutionResponse(BaseModel):
    """Code execution response"""

    success: bool
    stdout: str
    stderr: str
    error_message: Optional[str]
    execution_time_ms: Optional[int]
    environment_id: UUID

File: pymd_frontend/types/environment.ts
/**
 * Environment and Package types
 */

export interface Environment {
  id: string
  user_id: string
  name: string
  python_version: string
  volume_name: string
  status: 'active' | 'creating' | 'error' | 'deleted'
  total_size_bytes: number
  package_count: number
  last_used_at: string | null
  created_at: string
  updated_at: string
}

export interface EnvironmentCreate {
  name: string
  python_version?: string
}

export interface EnvironmentUpdate {
  name?: string
  python_version?: string
}

export interface EnvironmentListResponse {
  environments: Environment[]
  total: number
}

export interface EnvironmentStats {
  environment_count: number
  environment_limit: number
  total_size_bytes: number
  size_limit_bytes: number
  total_packages: number
  package_limit: number
  can_create_environment: boolean
}

export interface Package {
  id: string
  environment_id: string
  package_name: string
  version: string | null
  size_bytes: number | null
  installed_at: string
}

export interface PackageInstallRequest {
  packages: string[]
}

export interface PackageInstallResponse {
  success_count: number
  fail_count: number
  successful: Array<{
    package: string
    version: string | null
    message: string
  }>
  failed: Array<{
    package: string
    version: string | null
    message: string
  }>
}

export interface PackageListResponse {
  packages: Package[]
  total: number
  total_size_bytes: number
}

export interface RequirementsImportRequest {
  requirements: string
}

export interface RequirementsImportResponse {
  success_count: number
  fail_count: number
  successful: Array<{
    package: string
    version: string | null
    message: string
  }>
  failed: Array<{
    package: string
    version: string | null
    message: string
  }>
}
